test/distributed/org/apache/cassandra/distributed/test/BootstrapTest.java

```diff
@@ -18,85 +18,241 @@
+            }).drop();
+
+
+            Map<Integer, Long> connectionAttempts = new HashMap<>();
+            long deadline = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(10);
+
+            // Wait 10 seconds and check if there are any new connection attempts to the decomissioned node
+            while (System.currentTimeMillis() <= deadline)
+            {
+                for (int i = 2; i <= cluster.size(); i++)
+                {
+                    Object[][] res = cluster.get(i).executeInternal("SELECT active_connections, connection_attempts FROM system_views.internode_outbound WHERE address = '127.0.0.1' AND port = 7012");
+                    Assert.assertEquals(1, res.length);
+                    Assert.assertEquals(0L, ((Long) res[0][0]).longValue());
+                    long attempts = ((Long) res[0][1]).longValue();
+                    if (connectionAttempts.get(i) == null)
+                        connectionAttempts.put(i, attempts);
+                    else
+                        Assert.assertEquals(connectionAttempts.get(i), (Long) attempts);
+                }
+                LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(100));
+            }
+        }
     }
 
-    public void populate(ICluster cluster)
+    @Test
+    public void clientConnectionsDuringDecom() throws Throwable
     {
-        cluster.schemaChange("CREATE KEYSPACE " + KEYSPACE + " WITH replication = {'class': 'SimpleStrategy', 'replication_factor': " + 3 + "};");
-        cluster.schemaChange("CREATE TABLE " + KEYSPACE + ".tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck))");
+        ExecutorService executor = Executors.newFixedThreadPool(1);
+        try (Cluster cluster = builder().withNodes(4)
+                                        .withConfig(config -> config.with(NETWORK, GOSSIP, NATIVE_PROTOCOL))
+                                        .start())
+        {
+            populate(cluster, 0, 100);
+
+            try (com.datastax.driver.core.Cluster client = com.datastax.driver.core.Cluster.builder().addContactPoint("127.0.0.1")
+                                                                                           .withLoadBalancingPolicy(new RoundRobinPolicy())
+                                                                                           .build();
+                 Session session = client.connect())
+            {
+                CompletableFuture.runAsync(() -> {
+                    while (!Thread.currentThread().isInterrupted())
+                    {
+                        if (client.getMetadata().getAllHosts().size() == cluster.size())
+                            return;
+                        LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(100));
+                    }
+                }, executor).get(1, TimeUnit.MINUTES);
+
+                cluster.run(decomission(), 1);
+
+                Assert.assertEquals(cluster.size() - 1, client.getMetadata().getAllHosts().size());
```

Comment:
I ran this a few times locally and wasn't flaky, but curious how we know this will be stable?  I remember the client detecting the connection gets shut down, then runs a callback; what makes sure that the call back completes before this line?

Reply:
Client runs `onDown` on disconnect. So I think this should be fine, given we know connection was closed.